File: mo_lucro_app/lib/services/transaction_service.dart
import 'package:flutter/foundation.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/transaction_model.dart';

class TransactionService {
  final _db = Supabase.instance.client;

  String get _userId {
    final user = _db.auth.currentUser;
    if (user == null) throw Exception('Usuário não autenticado');
    return user.id;
  }

  Future<List<TransactionModel>> getTransactions() async {
    // Order by created_at — the live table doesn't have a 'date' column
    final res = await _db
        .from('transactions')
        .select()
        .eq('user_id', _userId)
        .order('created_at', ascending: false);

    return (res as List)
        .map((row) => TransactionModel.fromMap(row as Map<String, dynamic>))
        .toList();
  }

  Future<void> addTransaction({
    required String type,
    required double amount,
    required String category,
    required String description,
    required DateTime date,
  }) async {
    final payload = <String, dynamic>{
      'user_id': _userId,
      'type': type,
      'amount': amount,
      'category': category,
      'date': date.toIso8601String(),
    };

    if (description.isNotEmpty) {
      payload['description'] = description;
    }

    await _db.from('transactions').insert(payload);
  }

  Future<void> deleteTransaction(String id) async {
    await _db.from('transactions').delete().eq('id', id);
  }

  /// Returns total income and expense for the current month.
  /// Uses 'created_at' as the date filter since the live table may not have
  /// a separate 'date' column.
  Future<Map<String, double>> getMonthSummary(int year, int month) async {
    try {
      final start = DateTime(year, month, 1);
      final end = DateTime(year, month + 1, 1);

      final res = await _db
          .from('transactions')
          .select()
          .eq('user_id', _userId)
          .gte('created_at', start.toIso8601String())
          .lt('created_at', end.toIso8601String());

      double totalIncome = 0;
      double totalExpense = 0;

      for (final row in res as List) {
        final amount = (row['amount'] as num).toDouble();
        if (row['type'] == 'income') {
          totalIncome += amount;
        } else {
          totalExpense += amount;
        }
      }

      return {'income': totalIncome, 'expense': totalExpense};
    } catch (e) {
      // Never let a transaction query failure block the operations load
      debugPrint('[TransactionService.getMonthSummary] error: $e');
      return {'income': 0, 'expense': 0};
    }
  }
}

File: mobile/src/services/transactionService.ts
import { transactionFromRow, type Transaction, type TransactionType } from '@/models';
import { requireUserId, supabase } from '@/lib/supabase';

/** Porte de lib/services/transaction_service.dart. */

export async function getTransactions(): Promise<Transaction[]> {
  const userId = await requireUserId();

  // Ordena por created_at — a tabela em produção não tem coluna 'date'.
  const { data, error } = await supabase
    .from('transactions')
    .select()
    .eq('user_id', userId)
    .order('created_at', { ascending: false });

  if (error) throw error;
  return (data ?? []).map(transactionFromRow);
}

export async function addTransaction(params: {
  type: TransactionType;
  amount: number;
  category: string;
  description: string;
  date: Date;
}): Promise<void> {
  const userId = await requireUserId();

  const payload: Record<string, any> = {
    user_id: userId,
    type: params.type,
    amount: params.amount,
    category: params.category,
    date: params.date.toISOString(),
  };

  if (params.description.length > 0) {
    payload.description = params.description;
  }

  const { error } = await supabase.from('transactions').insert(payload);
  if (error) throw error;
}

export async function deleteTransaction(id: string): Promise<void> {
  const { error } = await supabase.from('transactions').delete().eq('id', id);
  if (error) throw error;
}

export type MonthSummary = { income: number; expense: number };

/**
 * Totais de entrada e saída do mês. Filtra por created_at porque a tabela em
 * produção pode não ter uma coluna 'date' separada.
 *
 * Nunca propaga erro: uma falha aqui não pode travar o carregamento das operações.
 */
export async function getMonthSummary(year: number, month: number): Promise<MonthSummary> {
  try {
    const userId = await requireUserId();

    const start = new Date(year, month - 1, 1);
    const end = new Date(year, month, 1);

    const { data, error } = await supabase
      .from('transactions')
      .select()
      .eq('user_id', userId)
      .gte('created_at', start.toISOString())
      .lt('created_at', end.toISOString());

    if (error) throw error;

    let income = 0;
    let expense = 0;

    for (const row of data ?? []) {
      const amount = Number(row.amount);
      if (row.type === 'income') income += amount;
      else expense += amount;
    }

    return { income, expense };
  } catch (error) {
    console.warn('[transactionService.getMonthSummary]', error);
    return { income: 0, expense: 0 };
  }
}
